### PR TITLE
LPS-190017 Class DestinationConfiguration no longer uses Liferay’s own implementation

### PR DESCRIPTION
**Pull request by TamyrisTorres**
Ticket: https://liferay.atlassian.net/browse/LPS-190017

Hi Kevin!

For this check, we need to include ThreadPoolExecutor when instantiating CallerRunsPolicy.

### CODE REVIEW

**Comment by TamyrisTorres** [comment]
ci:test:sf

**Comment by TamyrisTorres** [comment]
ci:test:relevant

**Comment by kevhlee** [comment]
Taking a look.

File: modules/util/source-formatter/src/main/java/com/liferay/source/formatter/check/UpgradeJavaThreadPoolExecutorClassCheck.java [hunk 1]
```diff
@@ -0,0 +1,87 @@
...
+		if (replaced) {
+			newContent = JavaSourceUtil.addImports(
+				newContent, "java.util.concurrent.ThreadPoolExecutor");
+		}
+
+		return newContent;
+	}
+
+	private static final Pattern _callerRunsPolicyPattern = Pattern.compile(
+		"\\w+\\s*\\w+\\s*[^)]\\s*(\\s*new\\s*CallerRunsPolicy\\s*\\(\\))");
```
Comment (kevhlee) [inline modules/util/source-formatter/src/main/java/com/liferay/source/formatter/check/UpgradeJavaThreadPoolExecutorClassCheck.java:85]: We could change this regex to match `RejectedExecutionHandler` variable declarations. That way, you do not have to extract the method call and check the class name.

Comment (kevhlee) [reply modules/util/source-formatter/src/main/java/com/liferay/source/formatter/check/UpgradeJavaThreadPoolExecutorClassCheck.java:85]: Also shouldn't the first `\\s*` be `\\s+` instead? `\\s+` ensures that at least 1 whitespace character is matched.

**Review by kevhlee — CHANGES_REQUESTED** [review]
Hi @TamyrisTorres, I am a little confused about what pattern is being automated here and what problem it is solving here. The ticket talks about `DestinationConfiguration`, yet the configuration class is not involved in any way in this check.

Also, just checking `RejectedExecutionHandler` variable declarations would not prevent `CallerRunsPolicy` being used as a `RejectedExecutionHandler`. For example, in our own code, we have usages like this:

```java
ThreadPoolExecutor threadPoolExecutor = new ThreadPoolExecutor(
	1, 2, TestUtil.KEEPALIVE_TIME, TimeUnit.MILLISECONDS, true, 3,
	new CallerRunsPolicy(), Executors.defaultThreadFactory(),
	markerThreadPoolHandler);
```

Can you elaborate on what problem this check solves?

**Comment by nicolas-sss** [comment]
> Hi @TamyrisTorres, I am a little confused about what pattern is being automated here and what problem it is solving here. The ticket talks about `DestinationConfiguration`, yet the configuration class is not involved in any way in this check.
> 
> Also, just checking `RejectedExecutionHandler` variable declarations would not prevent `CallerRunsPolicy` being used as a `RejectedExecutionHandler`. For example, in our own code, we have usages like this:
> 
> ```java
> ThreadPoolExecutor threadPoolExecutor = new ThreadPoolExecutor(
> 	1, 2, TestUtil.KEEPALIVE_TIME, TimeUnit.MILLISECONDS, true, 3,
> 	new CallerRunsPolicy(), Executors.defaultThreadFactory(),
> 	markerThreadPoolHandler);
> ```
> 
> It is not clear from the test inputs what problem this check is trying to solve. Can you elaborate on the exact problem?

Hi @kevhlee, the pattern for this check isn't properly about the DestinationConfiguration, sorry about that. It's about this pattern:

Before: 

```java
import com.liferay.portal.kernel.concurrent.CallerRunsPolicy;
import com.liferay.portal.kernel.concurrent.RejectedExecutionHandler;
import com.liferay.portal.kernel.concurrent.ThreadPoolExecutor;

RejectedExecutionHandler rejectedExecutionHandler =
				new CallerRunsPolicy();
```

Expected:
```java
import java.util.concurrent.RejectedExecutionHandler;
import java.util.concurrent.ThreadPoolExecutor;

RejectedExecutionHandler rejectedExecutionHandler =
				new ThreadPoolExecutor.CallerRunsPolicy();
```

The imports changed and the declaration should not use `new CallerRunsPolicy()` anymore.

**Comment by kevhlee** [comment]
Thanks for the explanation, @nicolas-sss. However, that only cleared things up for me only a little bit. To me, this check does not do much because it only tackles variable declarations, as I mentioned in my other comments. Why just variable declarations? Why not ___any usages___ of `com.liferay.portal.kernel.concurrent.CallerRunsPolicy`?

**Comment by nicolas-sss** [comment]
> Thanks for the explanation, @nicolas-sss. However, that only cleared things up for me only a little bit. To me, this check does not do much because it only tackles variable declarations. Why just variable declarations? Why not _**any usage**_ of `com.liferay.portal.kernel.concurrent.CallerRunsPolicy`? It does not handle the input case that I showed in my other comment. It is still unclear how this check is useful.
> 
> Also, this check only works if `ThreadPoolExecutor` is also imported. But what if `com.liferay.portal.kernel.concurrent.CallerRunsPolicy` is used without a `ThreadPoolExecutor`? What do we do then?

Hi @kevhlee,  the pattern is about the change on `RejectedExecutionHandler` use. Because of this, it is only in the variable declaration, the Liferay class `com.liferay.portal.kernel.concurrent.CallerRunsPolicy` still exists and is used the same in its previous input, in its previous input it does not need to be changed, only when related directly to the `RejectedExecutionHandler`. 
Previously Liferay classes used the 
`com.liferay.portal.kernel.concurrent.RejectedExecutionHandler` class as a variable that used the `com.liferay.portal.kernel.concurrent.CallerRunsPolicy`. But now is to switch to java.util.concurrent.RejectedExecutionHandler which uses the callers returned always by `java.util.concurrent.ThreadPoolExecutor`.

So after replacing import com.liferay.portal.kernel.concurrent.RejectedExecutionHandler by java.util.concurrent.RejectedExecutionHandler. We change the declaration from this:

```java
RejectedExecutionHandler rejectedExecutionHandler =
				new CallerRunsPolicy();
```

to this:

```java
RejectedExecutionHandler rejectedExecutionHandler =
			new ThreadPoolExecutor.CallerRunsPolicy() {...}
```

File: modules/util/source-formatter/src/main/java/com/liferay/source/formatter/check/UpgradeJavaThreadPoolExecutorClassCheck.java [hunk 2]
```diff
@@ -0,0 +1,87 @@
...
+
+		if (!fileName.endsWith(".java")) {
+			return content;
+		}
+
+		List<String> importNames = JavaSourceUtil.getImportNames(content);
+
+		if (!importNames.contains(
+				"com.liferay.portal.kernel.concurrent." +
+					"RejectExecutionHandler") &&
```
Comment (kevhlee) [inline modules/util/source-formatter/src/main/java/com/liferay/source/formatter/check/UpgradeJavaThreadPoolExecutorClassCheck.java:43]: Should be `RejectedExecutionHandler`.

**Review by kevhlee — CHANGES_REQUESTED** [review]

File: modules/util/source-formatter/src/main/java/com/liferay/source/formatter/check/UpgradeJavaThreadPoolExecutorClassCheck.java (hunk 2 above)
Comment (kevhlee) [inline modules/util/source-formatter/src/main/java/com/liferay/source/formatter/check/UpgradeJavaThreadPoolExecutorClassCheck.java:43]: This should be `RejectedExecutionHandler`.

File: modules/util/source-formatter/src/main/java/com/liferay/source/formatter/check/UpgradeJavaThreadPoolExecutorClassCheck.java [hunk 3]
```diff
@@ -0,0 +1,87 @@
...
+				StringUtil.replace(
+					methodCall, matcher.group(1),
+					"new ThreadPoolExecutor.CallerRunsPolicy()"));
+
+			replaced = true;
+		}
+
+		if (replaced) {
+			newContent = JavaSourceUtil.addImports(
+				newContent, "java.util.concurrent.ThreadPoolExecutor");
```
Comment (kevhlee) [inline modules/util/source-formatter/src/main/java/com/liferay/source/formatter/check/UpgradeJavaThreadPoolExecutorClassCheck.java:78]: Hey @nicolas-sss, we discussed this earlier. As you suggested, we need to also change the import `com.liferay.portal.kernel.concurrent.RejectedExecutionHandler` to `java.util.concurrent.RejectedExecutionHandler`.

**Review by kevhlee — CHANGES_REQUESTED** [review]

**Comment by nicolas-sss** [comment]
Resent at https://github.com/kevhlee/liferay-portal/pull/123

**Comment by nicolas-sss** [comment]
@kevhlee please close this Pr and review the new one

**Pull request closed without merging**